### DISCUSSION TITLE
multiple suggester?

### DISCUSSION THREAD

**Discussion by Cryptodarktrader** [SilentVoid13/Templater#1237] [Help]
I am trying to create a code that will open one suggestion and depending on the choices it will open others, and then open one more suggestion to ask if I want to add something else.

At the moment I have something like this:

```
<%* 
const categorie_tag = await tp.system.suggester([
"🥣Soup",
"❄️Cold",
"💚Vegan" 
], [
"🍽️Recipes/🍝Categories/🥣Soup",
"🍽️Recipes/🍝Categories/❄️Cold",
"🍽️Recipes/🍝Categories/💚Vegan" 
], false, "What type of Recipe is it?") 
%>
<% categorie_tag %>

<%* 
//🔽 Suggester that asks if we want to add more categories
const additional_category = await tp.system.suggester([
    "✅ Si", 
    "⛔ No" 
], [
    "✅ Si", 
    "⛔ No" 
], false, "Elegir opción:");
if (additional_category === "✅ Si") 
{ %>, <% tp.file.include("[[scr.categories]]") %><%* } 
else { %><%* } 
_%>
```

After choosing the category, it asks me if I want to add one more category and adds it and starts over.

On the other hand I have tried this code:
```
<%*
const categorie_tag = await tp.system.suggester([
  "🥣Soup",
  "❄️Cold",
  "💚Vegan" 
], [
  "🍽️Recipes/🍝Categories/🥣Soup",
  "🍽️Recipes/🍝Categories/❄️Cold",
  "🍽️Recipes/🍝Categories/💚Vegan"
], false, "What type of Recipe is it?") 

if (categorie_tag === "🍽️Recipes/🍝Categories/🥣Soup") {
  const soupType = await tp.system.suggester([
    "🍲Tomato Soup",
    "🥣Chicken Noodle Soup",
    "🍵Miso Soup"
  ], [
    "🍽️Recipes/🍝Categories/🥣Soup/🍲Tomato Soup",
    "🍽️Recipes/🍝Categories/🥣Soup/🥣Chicken Noodle Soup",
    "🍽️Recipes/🍝Categories/🥣Soup/🍵Miso Soup"

], false, "What type of Soup is it?") 
  return soupType;
} else {
  return categorie_tag;
}
%>
```

and it also works by opening a second suggester to ask me what type of soup if I choose soup and it returns the chosen soup.

My problem is when I try to merge them, that is to say that in addition to asking me the options after choosing a category (for example types of soup after choosing soup) it also then runs the suggester asking me if I want to add another category and if so I relaunch it. the code again.

Every time I try, it ignores the second block of code, if I add the yes or no suggester after the soup type suggester, the yes or no suggester does not execute and if I put it the other way around, it is the soup suggester that does not execute. . Can templater only execute an `iff` in the same code?

**Comment by Zachatoo** [comment]
Instead of using `tp.file.include()` to rerun the suggester, I'd recommend using a JS loop, like I'm doing [here](https://zachyoung.dev/posts/templater-multi-select-suggester), should make it much easier to merge your two pieces of code.

**Reply by Cryptodarktrader** [reply]
I understand, I thank you for the answer, it has been useful for me to investigate other ways of doing it, my problem comes with your code when I try to make the displayed value and the returned value be different and both be stored in variables.

The idea is that it shows, for example, "soup" and if I choose soup, that is stored in a variable that will go to the template on the one hand and also return the tag, for example "🍽️Recipes/🍝Categories/🥣Soup" and this is stored in another variable that will then go to the properties

That is a more or less controlled part, but if I try to also, when choosing the category, another suggester is displayed with more options in this example different types of soup, and these responses are added to the same variables.

**Reply by Zachatoo** [reply]
I had to make a change to my multiSuggester to get it to work with an array for the `textItems` argument, I almost always use a callback function instead of an array, forgot to test it with that.

<https://zachyoung.dev/posts/templater-multi-select-suggester#user-script>

With that update, here's how you could use my multiSuggester to prompt for multiple tags, and if one of your selected tags was a soup, it'll prompt for a specific type of soup.

```js
<%*
// Get list of tags, multi-select
const categorie_tags = await tp.user.multiSuggester(tp, [
  "🥣Soup",
  "❄️Cold",
  "💚Vegan",
], [
  "🍽️Recipes/🍝Categories/🥣Soup",
  "🍽️Recipes/🍝Categories/❄️Cold",
  "🍽️Recipes/🍝Categories/💚Vegan"
], false, "What type of Recipe is it?")
// If soup tag was selected as one of the tags, ask what type of soup
const soupCategorieIndex = categorie_tags.indexOf("🍽️Recipes/🍝Categories/🥣Soup");
if (soupCategorieIndex >= 0) {
  const soupType = await tp.system.suggester([
    "🍲Tomato Soup",
    "🥣Chicken Noodle Soup",
    "🍵Miso Soup",
  ], [
    "🍽️Recipes/🍝Categories/🥣Soup/🍲Tomato Soup",
    "🍽️Recipes/🍝Categories/🥣Soup/🥣Chicken Noodle Soup",
    "🍽️Recipes/🍝Categories/🥣Soup/🍵Miso Soup",
  ], false, "What type of Soup is it?")
  // Replace soup tag with more specific soup tag
  categorie_tags[soupCategorieIndex] = soupType;
}
tR += categorie_tags;
%>
```

**Reply by Cryptodarktrader** [reply]
The problem I see is that multisugester tells me that it is not a function for obsidian. In the end I have solved it 90% in the following way:

I have this code in my recipe note:

```
//🔽 --------------- CATEGORIES --------------- 🔽
// Create the chain with the categories tags to separate
const categories_tags = await tp.file.include("[[sc.categories]]"); 
const textosSeparados = categories_tags.split(', ');

// Obtén el último elemento de cada subcadena
let categories = textosSeparados.map(texto => {
    const partes = texto.split('/');
    return partes[partes.length - 1].trim();
}).join(', ');

// Elimina la última coma y espacio si existe


if (categories.endsWith(', ')) {
    categories = categories.slice(0, -2);
}
//-------------------------------------------------
```
It invokes the sc.categories script (which I leave below) where the categories of basic recipes would be and which includes the code to invoke the subcategories, in this case the dessert category for which I also left the script)

These scripts return the result of the tags only (no matter how hard I tried, I couldn't find a functional code that would return the element shown in the suggester in one variable and the real element that is returned in another, and that would allow me to open other suggesters).

then the script returns the label, example: 🍽️Recipes/🍝Categories/🥣Soup" then, separates all the results of the string by labels, (in case more than one option is chosen) then separates each part of the label by the "/" and returns the last part which is what interests us, example: 🥣Soup and I join all these parts into a new string separated by ","

My only problem so far is that I can't get it to remove the last "," from the end of the string, otherwise it works fine.

[tp.recipes.md](https://github.com/SilentVoid13/Templater/files/13249207/tp.recipes.md)
[sc.categories.md](https://github.com/SilentVoid13/Templater/files/13249212/sc.categories.md)
[sc.comida_postres.md](https://github.com/SilentVoid13/Templater/files/13249215/sc.comida_postres.md)
